### PR TITLE
Azure Spring Apps sub-resources

### PR DESCRIPTION
**Pull request by chutch1122**
Adds default values to `defaultResourcesState.go` for Azure Spring Apps sub-resources

- Build Service Agent Pool (https://learn.microsoft.com/en-us/rest/api/azurespringapps/build-service-agent-pool)
- Config Server (https://learn.microsoft.com/en-us/rest/api/azurespringapps/config-servers)
- Monitoring Settings (https://learn.microsoft.com/en-us/rest/api/azurespringapps/monitoring-settings)

### CODE REVIEW

**Comment by chutch1122** [comment]
I think this is ready for review. One thing I did notice is that this repository doesn't seem to have a `changelog` command in the makefile? (ie: `make changelog`).

**Comment by chutch1122** [comment]
@danielrbradley 

> I see the reference links in the description - which part of the docs did you use to infer the default values?
I created an Azure Spring Apps instance in Azure and pulled the default values by making API requests manually. I'll write this up in a follow-up comment after lunch.

> There's been a change to the azure-rest-api-specs submodule: 021e5ab (24th Oct) -> 8d9e220 (20th Oct). I think we want to revert this as this would make us point to an older version. Guessing this might have come in due to rebasing. Can leave this until the last minute as we upgrade versions most nights.

Sounds good. I remember seeing a change there when rebasing. Any advice on how to safely revert that change? I'm not used to working with git submodules.

**Comment by chutch1122** [comment]
@danielrbradley 

Here's a gist with the write-up on the default values: https://gist.github.com/chutch1122/0e925605194a62849d62cbbd0570d1ca

**Comment by danielrbradley** [comment]
Thanks for the gist - that was very helpful 🙇 

Here's my summary what I'm currently understanding here, please correct me if I've misunderstood.

1. Add default mappings for resources which can't be deleted. The properties in the map are then used to set values back to in place of doing a delete.
2. Add a default version for newly available resources to expose them through the SDK.

I've done a quick rebase for you over here which I think incorporates only the intended changes - feel free to reset your own branch to this commit if the change looks correct to you: https://github.com/pulumi/pulumi-azure-native/compare/azure-spring-apps

Pending questions: 
1. Should we set the default version of `BuildServiceAgentPool` to the stable `2022-04-01` version - rather than the preview version as this should have longer-term support.
2. Should `appInsightsInstrumentationKey` be being set to nil rather than an empty string `""`?
3. There was no `configServer` properties shown in your `configServers` gist - should this be nil rather than an empty object or does this not work in practice?
4. It doesn't appear there's a 'natural default' for the Build Service Agent Pool size - I see you've selected `S1` - is it possible to just not select a default and a 'delete' would just not modify this value?

**Comment by chutch1122** [comment]
> 1. Should we set the default version of BuildServiceAgentPool to the stable 2022-04-01 version - rather than the preview version as this should have longer-term support.

I figured since most of the other resources use the `2022-01-01-preview` version that I'd keep it consistent. Maybe it would make sense to upgrade everything together to the latest stable version in a follow-up PR?

> 2. Should appInsightsInstrumentationKey be being set to nil rather than an empty string ""?

According to the [documentation](https://learn.microsoft.com/en-us/rest/api/azurespringapps/monitoring-settings/update-put?tabs=HTTP#monitoringsettingproperties), null or empty string should both work.

> 3. There was no configServer properties shown in your configServers gist - should this be nil rather than an empty object or does this not work in practice?

Both an empty object and `null` both seem to work when using the ["Config Servers - Update Put"](https://learn.microsoft.com/en-us/rest/api/azurespringapps/config-servers/update-put?tabs=HTTP) endpoint. However, using `null` instead of an empty object produces the same response from the ["Config Servers - Get"](https://learn.microsoft.com/en-us/rest/api/azurespringapps/config-servers/get?tabs=HTTP) endpoint as the original configuration so it's probably "more correct".

Basically it comes down to:
```
GET https://management.azure.com/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.AppPlatform/Spring/{asaServiceName}/configServers/default?api-version=2022-09-01-preview

{
	"properties": {
		"provisioningState": "Succeeded"
                "configServer": {} // This property is returned in the response when you PUT an empty object as configServer. It is omitted when you PUT null as configServer
	},
        // ...
}
```

> 4. It doesn't appear there's a 'natural default' for the Build Service Agent Pool size - I see you've selected S1 - is it possible to just not select a default and a 'delete' would just not modify this value?

S1 is the smallest size available for the agent pool so if we can't truly delete it I thought it would be logical to minimize cost. It'd be fine not to modify the value too, of course. I don't have a preference either way.

**Comment by chutch1122** [comment]
I've reset my branch and updated the config server to be `nil`

**Review by danielrbradley — APPROVED** [review]
Happy with all the reasoning and the submodule change now being reverted. This can merge if tests are passing.

File: provider/pkg/openapi/defaultResourcesState.go [hunk 1]
```diff
@@ -121,7 +121,26 @@ var defaultResourcesStateRaw = map[string]map[string]interface{}{
 		"subnetResourceId": "*", // This is going to be a resource ID, so we choose accept any value here.
 		"swiftSupported":   true,
 	},
-	"/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Web/sites/{name}/sourcecontrols/web":                        {},
+	"/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Web/sites/{name}/sourcecontrols/web": {},
+	"/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.AppPlatform/Spring/{serviceName}/monitoringSettings/default": {
+		"properties": map[string]interface{}{
+			"traceEnabled":                  false,
+			"appInsightsInstrumentationKey": "",
```
[the remark this answers is not in the document]
Comment (chutch1122) [reply provider/pkg/openapi/defaultResourcesState.go]: They both work according to the documentation but null is probably more correct (similar to the config server). I've made a commit that reflects this change.

**Comment by chutch1122** [comment]
Looks like there's an issue with the PR build. It says "error: working tree is not clean, aborting!" 

What needs to happen to get this fixed?

**Comment by thomas11** [comment]
> Looks like there's an issue with the PR build. It says "error: working tree is not clean, aborting!"

@chutch1122, that typically means that something else has changed in the meantime and the SDK needs to be regenerated. I'll take a look.

**Comment by viveklak** [comment]
@chutch1122 thanks so much for the PR. I pulled this into #2055 to avoid more churn on the PR (needed to run `make clean` on the PR to remove files for removed types etc.). Thanks very much for your contribution!

**Pull request closed without merging**